### PR TITLE
Added events for crop growth

### PR DESCRIPTION
**Pull request by Draco18s**
New block event: CropGrowEvent
Patches for BlockChorusFlower, BlockCocoa, BlockCrops, BlockNetherWart,
BlockReed, and BlockStem (all of the vanilla "growing" blocks).

Example mod:
https://github.com/Draco18s/ReasonableRealism/tree/Farming-Example

### CODE REVIEW

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (kashike) [inline patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: ...fix your formatting. You should not be changing the patch like this.

**Review by kashike — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: I didn't touch the patch files directly.  When I fixed the formatting I ran genPatches again and sync'ed the altered patch files.

Comment (LexManos) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: NEVER 'fix' formatting of vanilla code...

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java
Comment (liach) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: `@HasResult`

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (covers1624) [inline patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Don't use imports in patches. Use fully qualified names.

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: and you should post the event only after random check has passed.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Jesus christ, and here I thought I was trying to make sure I was using four spaces and not tabs and figured the only place there would be tabs was in front of any code _I_ had written.  I was trying to be particularly careful about avoiding altering the formatting of the vanilla code I was working around (as my IDE kept trying to "correct" things).

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Oh, and this particular section of code is wrapped in an extra indentation.  There's an additional if(){...} around the vanilla code, that's why it's tabbed in farther.  Should it not be?  The style guide makes no mention of this.

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Ok, thanks, but that might want to be added to the style guide. How would I have known?

Comment (KnightMiner) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: From [CONTRIBUTING.md](https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/CONTRIBUTING.md) (which appears at the top when making an issue/PR), there is a link leading [here](https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge#conventions-for-coding-patches-for-a-minecraft-class-javapatch) which mentions said style guideline.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: My apologies, it was late and I was frustrated and didn't see the "more info here" link.

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (covers1624) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Keep patches as small as possible, so don't indent it at all. just a brace on the same indent as the previous will be fine.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Ok thank you. I was trying to follow the "indent your code properly" convention as I saw nothing else about it (and still don't).

**Comment by Draco18s** [comment]
Ok, I think I have addressed the issues people had with things.  I am just not sure how to squash the commits on the branch as I don't want to rebase onto the master of my fork.

File: patches/minecraft/net/minecraft/block/BlockCactus.java.patch
Comment (iTitus) [inline patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: This seems to be the same paragraph as before, just indented.
Just leave it unindented and change only necessary lines.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: Thanks. I was sure I'd had done that (extra indentation) in another file other than BlockChorusFlower but when I was reviewing the java code, and then the generated patches, I didn't find it.

**Comment by williewillus** [comment]
is this meant to be general enough to where mods could fire this event themselves?

If so, I would make the comments more general to match (instead of enumerating the vanilla cases)

**Comment by Draco18s** [comment]
@williewillus Essentially yes.  I figure _most_ mod crops would be inheriting from one of those classes without needing to override (and unfortunately the logic can't be handled by BlockBush) but in the case that they don't they'd have to fire the event themselves. Suggested rephrasing?

**Comment by williewillus** [comment]
Just don't restrict it to vanilla.

Something like:

Fired when any ~~of the~~ "growing age" blocks (**such as cacti, chorus plants, or crops in vanilla**) attempt to advance to the next growth age state during a random tick.

**Comment by Draco18s** [comment]
Ok, thanks. Updating that now.

**Comment by evieluvsrainbows** [comment]
@Draco18s Don't worry about squashing your commits. @LexManos and the other members who have write access can automatically squash your commits via GitHub itself since GH added that ability a while ago.

**Comment by Draco18s** [comment]
@KamranMackey I asked because it was in the contributor's readme.
Which other parts thereof are out of date, such as "how to install" being flat wrong.

**Comment by LexManos** [comment]
The instruction for contributing to Forge have not changed in 3 years. Not sure why you've had such issues but its a simple process.

**Comment by Draco18s** [comment]
@LexManos as mentioned in the reply to the comment where you gave me working instructions... "the current readme doesn't say that."  Current info says "gradlew setup" and your comment was "gradlew _setupForge_" followed by _____ (it literally says "you're done") vs. "import existing project."

**Comment by LexManos** [comment]
No idea what the hell you're looking at, but the wiki is just fine: https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge Ah found where you might be looking, it still says setupForge. But the eclipse folder no longer exists. So meh.

**Comment by Draco18s** [comment]
@LexManos The... readme on the front page of the repo...?
https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/README.md
Which incidentally also does not contain a link to the wiki for more information.

File: patches/minecraft/net/minecraft/block/BlockCactus.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: This is where you would use ForgeHooks having this big block of code in each patch over and over is bad.

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: First off, spacing.
Also, a 'ForgeHooks.cropsPre(x,y,z, rand.nextInt(1)==0)' would make this code and a lot of other patches A LOT cleaner.

File: patches/minecraft/net/minecraft/block/BlockCocoa.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockCocoa.java.patch]: You tab this section of code in another layer which makes the patches bigger and makes it hard to tell what you add and what is vanilla, fix that.

File: patches/minecraft/net/minecraft/block/BlockCrops.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockCrops.java.patch]: See above comment about ForgeHooks.

File: patches/minecraft/net/minecraft/block/BlockNetherWart.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockNetherWart.java.patch:15]: ForgeHooks and unindent.

File: patches/minecraft/net/minecraft/block/BlockReed.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockReed.java.patch]: Again nasty patch changing indent and not making it clear whats vanilla and what your changes are.

File: patches/minecraft/net/minecraft/block/BlockStem.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockStem.java.patch:9]: again so much cleaner if you did: 
if ForgeHooks.growPre(x, y, z, rand.nextInt((int)(25.0F / f) + 1) == 0)

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: brace

**Comment by Draco18s** [comment]
Thanks for the comments, @LexManos I'll see about fixing those up.

**Comment by Draco18s** [comment]
@LexManos Alright, I did the ForgeHooks thing just for BlockChorusFlower.  Is that what you are expecting? Wanted to make sure I did it right.

File: patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Could probably make this smaller with ForgeHooks to.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Looks like it would be...shorter by 69 characters.
Do it, yes no?

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: @LexManos Didn't think I would need to include an @mention. Using ForgeHooks would save 69 characters for that line.  Do you want me to do that? (I don't know what the threshold for saving size is and am asking for direction, I'd really like to get this merged so I can start using it).

Comment (LexManos) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Didn't think I would have to answer a stupid question that I told you to do...
Does it make the code look better, and easier to understand, does it make the code less repetitious in the other dozen patches you do?
Then yes it probably should be cleaned up.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch]: Thanks @LexManos I am only trying to understand your wishes more clearly.  "Could probably make this smaller" isn't a very clear instruction (the sentence contains two conditional clauses).  "If this can be made smaller by using a ForgeHook, please do so" would have been more instructive.  Thank you for your input and I will update the other classes in this manner as well.

**Comment by Draco18s** [comment]
How about over here, what's holding this back?

**Comment by Draco18s** [comment]
Hey @LexManos any additional changes needed for this to get merged?  I've had no comments for a week+

File: patches/minecraft/net/minecraft/block/BlockCactus.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: Why delete the line?

File: patches/minecraft/net/minecraft/block/BlockCrops.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockCrops.java.patch:8]: change whitespace?

File: patches/minecraft/net/minecraft/block/BlockReed.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockReed.java.patch]: ...

File: patches/minecraft/net/minecraft/block/BlockStem.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockStem.java.patch]: ...

File: patches/minecraft/net/minecraft/block/BlockCactus.java.patch
Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: @LexManos I am doing my best to follow the "how to contribute" guidelines.  In this case here I put the new line of code onto an existing blank line.  If that causes a patch change that is undesirable, then the how-to wiki page needs to be updated to reflect that case.  It currently has the insert of a new statement between two lines with _no_ blank line between them stating "don't add blank lines."  I used one that was there, if that was wrong, then I have been given bad instructions.

Comment (LexManos) [reply patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: its about balancing the patch file size and making things look okay in the end code.
This for example is just dumb, there is no real reason to delete the whitepsace as it just ads extra onto the patch file.
And deleting the whitespace makes the code more of a clump and harder to read.

Making your patch file bigger just to futz with whitespace is bad, your changes should be JUST whats needed to make your changes work.

Comment (Draco18s) [reply patches/minecraft/net/minecraft/block/BlockCactus.java.patch]: @LexManos My point was that I was _attempting not to mess with the whitespace_ and the how to contribute page does not make it clear _how to do that_ so I made a best guess. If I did it wrong, that's fine, I'll fix it, but the instructions on what I did wrong don't exist.  I can fix this here and now, but that doesn't help someone else who's new.  Prior to starting this endeavor _I didn't know how to read patch files._  I'm slowly starting to understand what it is that you want, but please, rather than being snarky all you have to say is "this added that extra line that makes the patch files harder to read. It's cleaner if you only insert lines, even if there's already a blank one."

**Comment by Draco18s** [comment]
Ok, all patches cleaned of white space alterations per @LexManos instruction

**Comment by LexManos** [comment]
Jesus you still cant figure out how to do whitespace STOP IT.
But whatever -.-

**Pull request merged**
